internal/fans/nvidia.go

```diff
@@ -0,0 +1,228 @@
+	"errors"
+	"os"
+
+	"github.com/NVIDIA/go-nvml/pkg/nvml"
+	"github.com/markusressel/fan2go/internal/configuration"
+	"github.com/markusressel/fan2go/internal/ui"
+)
+
+type NvidiaFan struct {
+	Label             string                  `json:"label"`
+	Index             int                     `json:"index"`
+	RpmMovingAvg      float64                 `json:"rpmMovingAvg"`
+	Config            configuration.FanConfig `json:"config"`
+	MinPwm            *int                    `json:"minPwm"`
+	StartPwm          *int                    `json:"startPwm"`
+	MaxPwm            *int                    `json:"maxPwm"`
+	FanCurveData      *map[int]float64        `json:"fanCurveData"`
+	Rpm               int                     `json:"rpm"`
+	Pwm               int                     `json:"pwm"`
+	RunAtMaxSpeed     bool                    // to emulate PWM mode 0
+	SupportedFeatures int                     // (1 << FeaturePwmSensor) | (1 << FeatureRpmSensor) or similar
+	// TODO: probably SupportedFeatures doesn't have to be saved to config? set in GetDevices()
+	// TODO: put nvml.Device here? though in reality it should be shared
+	//   between all fans and sensors of that device (but multiple devices can exist
+	//   when the system has multiple GPUs)
+}
+
+// helper function to turn an nvml error/return code into a go error
+// (also handles success by returning nil)
+func nvError(errCode nvml.Return) error {
+	if errCode == nvml.SUCCESS {
+		return nil
+	}
+	return errors.New(nvml.ErrorString(errCode))
+}
+
+func (fan *NvidiaFan) getNvFanIndex() int {
+	// seems like fan indices in fan2go are 1-based,
+	// here we start at 0.
+	return fan.Config.Nvidia.Index - 1
+}
+
+func (fan *NvidiaFan) GetId() string {
+	return fan.Config.ID
+}
+
+func (fan *NvidiaFan) GetMinPwm() int {
+	// if the fan is never supposed to stop,
+	// use the lowest pwm value where the fan is still spinning
+	if fan.ShouldNeverStop() {
+		if fan.MinPwm != nil {
+			return *fan.MinPwm
+		}
+	}
+
+	return MinPwmValue
+}
+
+func (fan *NvidiaFan) SetMinPwm(pwm int, force bool) {
+	if fan.Config.MinPwm == nil || force {
+		fan.MinPwm = &pwm
+	}
+}
+
+func (fan *NvidiaFan) GetStartPwm() int {
+	if fan.StartPwm != nil {
+		return *fan.StartPwm
+	} else {
+		return 100
+	}
+}
+
+func (fan *NvidiaFan) SetStartPwm(pwm int, force bool) {
+	if fan.Config.StartPwm == nil || force {
+		fan.StartPwm = &pwm
+	}
+}
+
+func (fan *NvidiaFan) GetMaxPwm() int {
+	if fan.MaxPwm != nil {
+		return *fan.MaxPwm
+	} else {
+		return 100
+	}
+}
+
+func (fan *NvidiaFan) SetMaxPwm(pwm int, force bool) {
+	if fan.Config.MaxPwm == nil || force {
+		fan.MaxPwm = &pwm
+	}
+}
+
+func (fan *NvidiaFan) GetRpm() (int, error) {
+	return fan.GetPwm()
+	// TODO: for driver 565 and newer we can get the actual RPM with custom cgo
+	// code, see my_DeviceGetFanSpeedRPM() in Daniel's nvml test code.
+	// But otherwise, I guess returning the current PWM should also work as it
+	// at least allows creating a fan curve
```

Comment:
Hmm, I get the idea, however the values returned by `GetPwm` will not be in a range thats reasonable for RPM values, which would probably cause confusion among users. Contrary to the initial intentions of fn2go, the RPM curve is current **not used** to control a fans speed, so messing with this is not really beneficial atm.